src/Shared/SecretsStore.cs

```diff
@@ -49,39 +68,50 @@ public SecretsStore(string userSecretsId)
 
     public void Save()
     {
-        EnsureUserSecretsDirectory();
-
-        var contents = new JsonObject();
-        if (_secrets is not null)
+        var semaphore = GetLock(_userSecretsId);
+        semaphore.Wait();
+        try
         {
-            foreach (var secret in _secrets.AsEnumerable())
+            // Reload from disk to merge with any concurrent changes
+            var currentSecrets = Load(_secretsFilePath);
+            
+            // Merge our changes with what's on disk
+            foreach (var kvp in _secrets)
+            {
+                currentSecrets[kvp.Key] = kvp.Value;
```

Comment:
The keys that we are iterating over here are already flattened here so we don't have to worry about the hierarchical collision problem?